docs/reference/xpkg.md

```diff
@@ -0,0 +1,201 @@
+is only concerned with the layer descriptors array in an image manifest and does
+not impose additional requirements on any other portions of the manifest. The
+following requirements are imposed on the layer descriptors array:
+
+- One (1) layer descriptor in the array MAY have an [annotation] with key
+  `io.crossplane.xpkg` and value `base`.
+- Any number of layer descriptors in the array MAY have an annotation with key
+  `io.crossplane.xpkg` and arbitrary value. Whether multiple layer descriptors
+  may have the same value is left to the specification of the consumer of those
+  layers.
+
+> As evidenced by the fact that annotations are provided as a map of
+> _string-string_, no single descriptor will contain multiple
+> `io.crossplane.xpkg` annotations.
+
+Crossplane is only concerned with the layer with the `base` annotation, and any
+other layers with the `io.crossplane.xpkg` key are used to signify to
+third-party consumers that a layer contains content related to the `xpkg` that
+may be specific to a given consumer.
+
+If no layer descriptors have an annotation in the form `io.crossplane.xpkg:
+base`, the resultant filesystem from [applying changesets] from all layers will
+be used.
+
+> Crossplane prefers the usage of layers to define additive package content as
+> it provides a clean mechanism to build an `xpkg` through a series of stages. A
+> valid `xpkg` can be produced and later modified while verifying that the
+> integrity of the existing content is not violated, which ensures that
+> Crossplane's package manager will process the resulting `xpkg` in the same
+> manner as the it would prior to modification.
+
+### Configuration
+
+Crossplane imposes no additional requirements on image configuration and does
+not consider its contents when processing a package.
+
+### Layers
+
+As described above, Crossplane is only concerned with the single layer
+referenced by the descriptor containing `io.crossplane.xpkg: base` if
+distinguished. Crossplane imposes no additional restrictions on any other
+layers, including those with a `io.crossplane.xpkg` annotation but a value other
+than `base`, but does require the following of the `xpkg` base layer:
+
+- A single file with name `package.yaml` MUST exist in the root directory of the
+  `xpkg` base layer if distinguished, or in the root of the image filesystem
+  after all layer changesets are applied.
+- The `package.yaml` file MUST contain a valid [YAML stream].
+- All other content in either the `xpkg` base layer, or the full image
+  filesystem is ignored by Crossplane.
+
+> The ability to use the image's flattened filesystem is primarily for backwards
+> compatibility and is not encouraged, especially in the event that an image
+> contains more than just `xpkg` related content, due to the fact that
+> accidentally overwriting or modifying the `xpkg` layer contents in subsequent
+> layers when constructing an image could cause the package to be invalid.
```

Comment:
I strongly considered dropping this and instead making the "backwards compatibility" requirement that the image only consists of a single (1) layer. This would technically still be a breaking change because today we will honor a package with multiple layers and a `package.yaml` defined at any layer, but in practice, every meaningful `xpkg` we are aware of just includes the single layer. So while it would be a breaking change, it is unlikely to have negative consequences and the existing behavior could be considered "unspecified" rather than "explicitly supported". The real benefit of using the single layer for backwards compatibility is that third-party consumers who wanted to honor the spec for backwards compatibility would still only have to download a single layer for parsing.

Reply:
I feel slightly better about what is proposed here - while I take your point that in practice pretty much everything is today a single layer I'd rather optimise for broader compatibility here (e.g. supporting theoretical existing packages w/multiple layers) vs optimising for guaranteeing the ability to always download only one layer.

Reply:
I agree with keeping the compatibility. A lot of tools and some registries do assume as if the image is just a file blob, which is fair IMO.